src/textinput/tests/functional/TextInput.ts

```diff
@@ -0,0 +1,143 @@
+import * as registerSuite from 'intern!object';
+import * as assert from 'intern/chai!assert';
+import * as css from '../../styles/textinput.m.css';
+import * as baseCss from '../../../common/styles/base.m.css';
+import * as keys from 'leadfoot/keys';
+
+function getPage(remote: any) {
+	return remote
+		.get('http://localhost:9000/_build/common/example/?module=textinput')
+		.setFindTimeout(5000);
+}
+
+registerSuite({
+	name: 'TextInput',
+	'TextInput should be visible'(this: any) {
+		return getPage(this.remote)
+			.findByCssSelector(`#example-text .${css.root}`)
+			.isDisplayed()
+			.findByCssSelector(`.${css.inputWrapper}`)
+			.isDisplayed()
+			.end()
+
+			.findByCssSelector(`.${css.input}`)
+			.getSize()
+			.then(({ height, width }: { height: number; width: number; }) => {
+				assert.isAbove(height, 0, 'The height of the input should be greater than zero.');
+				assert.isAbove(width, 0, 'The width of the input should be greater than zero.');
+			})
+			.end()
+			.end();
+	},
+	'TextInput label should be as defined'(this: any) {
+		return getPage(this.remote)
+			.findByCssSelector(`#example-text .${css.root}`)
+			.getVisibleText()
+			.then((text: string) => {
+				assert.strictEqual(text, 'Name');
+			});
+	},
+	'TextInput should gain focus when clicking on the label'(this: any) {
+		let elementClassName: string;
+		return getPage(this.remote)
+			.findByCssSelector(`#example-text .${css.root}`)
+			.findByCssSelector(`.${css.input}`)
+			.getProperty('className')
+			.then(function(className: string) {
+				elementClassName = className;
+			})
+			.end()
+			.click()
+			.getActiveElement()
+			.then(function(element: any) {
+				element
+					.getProperty('className')
+					.then(function(className: string) {
+						assert.strictEqual(className, elementClassName);
+					});
+			})
+			.end();
+	},
+	'TextInput should allow input to be typed'(this: any) {
+		const testInput = 'test text';
+		return getPage(this.remote)
+			.findByCssSelector(`#example-hidden-label .${css.root}`)
+			.click()
+				.findByCssSelector(`.${css.input}`)
+				.type(testInput)
+				.getProperty('value')
+				.then((value: string) => {
+					assert.strictEqual(value, testInput);
+				})
+				.end()
+			.end();
+	},
+	'Hidden label should not be displayed'(this: any) {
+		return getPage(this.remote)
+			.findByCssSelector(`#example-hidden-label .${css.root}`)
+			.getVisibleText()
+			.then((text: string) => {
+				assert.isTrue(text && text.length > 0);
+			})
+				.findByCssSelector(`.${baseCss.visuallyHidden}`)
+				.getSize()
+				.then(({ height, width }: { height: number; width: number; }) => {
+					assert.isAtMost(height, 1, 'The label text height should be no more than 1px.');
```

Comment:
Do you think this is the best way to perform this test? It feels like it's testing the inner workings of label, also that it's testing the implementation of the `visuallyHidden` class. I'd have thought that testing for the presence of this class alone is enough?